modules/server/src/completion.ts

```diff
@@ -0,0 +1,339 @@
+      const line = lines[i].trim()
+      if (line.match(/^(public|private|protected|internal)?\s*(class|interface|enhancement|enum|structure)\s+/)) {
+        return false
+      }
+    }
+    return currentLine < 3 // More restrictive: only first 3 lines are "file start"
+  }
+
+  /**
+   * Check if position is inside a class body (includes enhancement bodies)
+   */
+  private isInClassBody(lines: string[], currentLine: number): boolean {
+    let braceDepth = 0
+    let inClassLike = false
+    
+    for (let i = 0; i <= currentLine && i < lines.length; i++) {
+      const line = lines[i].trim()
+      
+      // Check for class-like declarations (class, interface, enhancement, etc.)
+      if (line.match(/^(public|private|protected|internal)?\s*(class|interface|enhancement|enum|structure)\s+/)) {
+        inClassLike = true
+      }
+      
+      // Count braces
+      const openBraces = (line.match(/\{/g) || []).length
+      const closeBraces = (line.match(/\}/g) || []).length
+      braceDepth += openBraces - closeBraces
+    }
+    
+    return inClassLike && braceDepth > 0
+  }
+
+  /**
+   * Check if position is inside a function body
+   */
+  private isInFunctionBody(lines: string[], currentLine: number): boolean {
+    let classLikeBraceDepth = 0
+    let functionBraceDepth = 0
+    let inFunction = false
+    let inClassLike = false
+    
+    for (let i = 0; i <= currentLine && i < lines.length; i++) {
+      const line = lines[i].trim()
+      
+      // Check for class-like declaration
+      if (line.match(/^(public|private|protected|internal)?\s*(class|interface|enhancement|enum|structure)\s+/)) {
+        inClassLike = true
+        continue
+      }
+      
+      // Check for function declaration (only count if we're inside a class-like construct)
+      if (inClassLike && line.match(/(function|construct)\s+/)) {
+        inFunction = true
+        // Reset function brace depth when we encounter a new function
+        functionBraceDepth = 0
+      }
+      
+      // Count braces
+      const openBraces = (line.match(/\{/g) || []).length
+      const closeBraces = (line.match(/\}/g) || []).length
+      
+      // Track class-like braces
+      if (inClassLike) {
+        classLikeBraceDepth += openBraces - closeBraces
+      }
+      
+      // Track function braces if we're in a function
+      if (inFunction && classLikeBraceDepth > 0) {
+        functionBraceDepth += openBraces - closeBraces
+        
+        // If we've closed the function, we're no longer in it
+        if (functionBraceDepth <= 0) {
+          inFunction = false
+        }
+      }
+    }
+    
+    return inClassLike && inFunction && classLikeBraceDepth > 0 && functionBraceDepth > 0
+  }
+
+  /**
+   * Get keywords relevant for the current context
+   */
+  private getKeywordsForContext(context: CompletionContext): GosuKeywordInfo[] {
+    const keywords: GosuKeywordInfo[] = []
+    
+    if (context.isFileStart) {
```

Comment:
Required unit tests that will confirm that we receive the exact keywords as in the G4 grammar defined, to avoid the issue of "grammar modification is not reflected in code"